### DISCUSSION TITLE
Video and image load time in flutter and supabase

### DISCUSSION THREAD

**Discussion by Gourishh** [community/community#127663] [Programming Help]
### Body

I am making a social media app using flutter and supabase. I am trying to replicate yt shorts/insta reels into my app but the difference is in the content type, I could be a video, image or audio. The main problem I am facing is the load time. It takes atleast 5-10 secs to load any content which is not good for UX experience. Can anybody guide me through the best method to implement this?

I tried compression but it didn't made a big difference. I have written basic code for video_player and image.network widget. My content is stored in supabase storage.

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by Hadikamali** [comment]
Hi @Gourishh 

To improve the loading time of media content in a Flutter application using Supabase storage, other things need to be done. Here are a few things:
1-CDN
2. Efficient Media Loading
3. Media Preloading
4. Lazy Loading
5. Caching
6. Optimize Media Files
7. Asynchronous Loading
8. Supabase Configuration
9. Flutter Performance Best Practices
Follow Flutter’s best practices for performance:
- Avoid excessive rebuilds by using `const` widgets where possible.
- Use `ListView.builder` for lists to ensure that only the visible items are built.

### Example Code for Image Preloading and Caching

Here is a simple example of using the `cached_network_image` package for caching and preloading images:

```dart
import 'package:flutter/material.dart';
import 'package:cached_network_image/cached_network_image.dart';

class MediaContent extends StatelessWidget {
  final String imageUrl;

  MediaContent({required this.imageUrl});

  @override
  Widget build(BuildContext context) {
    return CachedNetworkImage(
      imageUrl: imageUrl,
      placeholder: (context, url) => CircularProgressIndicator(),
      errorWidget: (context, url, error) => Icon(Icons.error),
    );
  }
}
```

### Example Code for Video Preloading

For video preloading, you can use the `video_player` package and control buffering manually:

```dart
import 'package:flutter/material.dart';
import 'package:video_player/video_player.dart';

class VideoContent extends StatefulWidget {
  final String videoUrl;

  VideoContent({required this.videoUrl});

  @override
  _VideoContentState createState() => _VideoContentState();
}

class _VideoContentState extends State<VideoContent> {
  late VideoPlayerController _controller;

  @override
  void initState() {
    super.initState();
    _controller = VideoPlayerController.network(widget.videoUrl)
      ..initialize().then((_) {
        setState(() {});
      });
  }

  @override
  Widget build(BuildContext context) {
    return _controller.value.isInitialized
        ? AspectRatio(
            aspectRatio: _controller.value.aspectRatio,
            child: VideoPlayer(_controller),
          )
        : Center(child: CircularProgressIndicator());
  }

  @override
  void dispose() {
    super.dispose();
    _controller.dispose();
  }
}
```
Powered by ChatGPT

**Reply by ebndev** [reply]
Hi @Hadikamali, thanks for participating in GitHub Discussions!

Please note: we’ve clarified our stance on using generative AI tools like ChatGPT within our Community via [this announcement](https://github.com/orgs/community/discussions/69006). Please review the guidelines to ensure your post meets them as failure to adhere to those rules can result in action taken by our moderator team.  

You can read our updated [Code of Conduct](https://github.com/community/community/blob/baa5cd9020ef3f2bbf86dca1191d9c90f9b0822d/CODE_OF_CONDUCT.md#reasonable-use-of-ai-generated-content) and the [announcement](https://github.com/orgs/community/discussions/69006) for more details. Thank you for helping us maintain an authentic and beneficial space for everyone.

**Comment by ebndev** [comment]
Hey @Gourishh, thanks for being a part of the GitHub Community!

Check out this previous [Discussion](https://github.com/orgs/supabase/discussions/3778#discussioncomment-1594881) where an answer suggests storing the images in a public URL, instead of fetching the image each time. 

If you are already doing this, you could also try caching the images but it would only improve reload time and not when the user first visits the page. 😕 That might help you dig into other possible solutions though! Here's the documentation on [Supabase Storage CDN](https://supabase.com/docs/guides/storage/cdn/fundamentals).

**Comment by Gourishh** [comment]
I have found solution through compression. It's not 100% what I wanted but it is very effective so I am marking this post as answer. I used package - [video_compress](https://pub.dev/packages/video_compress) for compressing videos. Still finding a good one for image.

**Reply by ebndev** [reply]
Nice!
